src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -375,32 +373,40 @@ bool ObjectMonitor::enter_for(JavaThread* locking_thread) {
     return false;
   }
 
-  enter_for_with_contention_mark(locking_thread, contention_mark);
+  bool success = TryLockWithContentionMark(locking_thread, contention_mark);
+
+  assert(success, "Failed to enter_for: locking_thread=" INTPTR_FORMAT
+         ", this=" INTPTR_FORMAT "{owner=" INTPTR_FORMAT "}",
+         p2i(locking_thread), p2i(this), p2i(owner_raw()));
   assert(owner_raw() == locking_thread, "must be");
   return true;
 }
 
-bool ObjectMonitor::try_enter(JavaThread* current) {
-  // TryLock avoids the CAS
+bool ObjectMonitor::try_enter(JavaThread* current, bool check_owner) {
```

Comment:
The `check_owner` name is a little confusing to me. To me it looks more like `check_for_recursion` or `handle_recursion`.

Reply:
I think the name should either describe what setting the value actually does, but if it is just a hack to do what the comment bellow says, then it sounds like a friend declaration for `SharedRuntime::monitor_exit_helper()` is what is wanted. (Or make TryLock() public.)
> Set check_owner to false (it's default value is true) if you want
> to use ObjectMonitor::try_enter() as a public way of doing TryLock().
> Used this way in SharedRuntime::monitor_exit_helper().

Reply:
fixed